### ISSUE TITLE
The error message `Hiding … has no effect` should be improved

### ISSUE THREAD

**Issue by mietek** [agda/agda#3928]
As seen in https://github.com/agda/agda/issues/3879#issuecomment-512468153.

Attempting to write
```agda
open import Data.Fin hiding (0F) using (Fin ; zero ; suc)
```
results in the following error message:
```
Hiding 0F has no effect
when scope checking the declaration
  open import Data.Fin using (Fin; zero; suc) hiding (0F)
```

This is misleading.  There are, in fact, two different methods to hide `0F` that have different effects.

### Method 1.
The following code fragments both make all identifiers in `Data.Fin` accessible unqualified, except for `Data.Fin.0F`, which is made **accessible fully-qualified**:
```agda
open import Data.Fin hiding (0F)
```
and
```agda
import Data.Fin
open Data.Fin hiding (0F)
```

### Method 2.
The following code makes all identifiers in `Data.Fin` accessible unqualified, except for `Data.Fin.0F`, which is made **inaccessible**:
```agda
import Data.Fin hiding (0F)
open Data.Fin
```

**Comment by Blaisorblade** [comment]
How can we improve this?

I’d propose to say “0F is already hidden by `using` clause”. That’s why hiding has no effect. There, ‘hidden’ refers to method 1. But I see that doesn’t address your point.

But I don’t see how to address it. Currently, `open import` statements can only use “method 1”, and `import` statements can only use method 2. Would you want more complex rules? That would seem confusing... Maybe method 2 could use another clause.

**Comment by mietek** [comment]
I expected
```agda
open import Data.Fin hiding (0F) using (Fin ; zero ; suc)
```
to work as method 2.  If we were to make it work as method 2, then we would get rid of the error condition, and the error message would be dropped.

Otherwise, the two meanings of `hiding` should first be clarified in the documentation (https://github.com/agda/agda/issues/3927).  Then, the error message could use the same wording as the documentation.

**Comment by Blaisorblade** [comment]
TLDR: the distinct meanings of `hiding` are clearly confusing and is bound to cause confusing expectations, so making this less arcane requires either docs or an alternative syntax for meaning 2, preferably not involving a new keyword. Alternatively, one could address #3879 so that `import Foo hiding (bar)` becomes much less necessary (which might be simplest) — so this issue would become less important.

The basic problem is that one could already guess that `import Foo hiding (bar); open Foo` is equivalent to `open import Foo hiding (bar)`, and it isn’t. As a special case, one could indeed have expected `import Data.Fin hiding (0F); open Data.Fin using (Fin ; zero ; suc)` to be equivalent to `open import Data.Fin hiding (0F) using (Fin ; zero ; suc)`. Technically, that’s even an option, tho making `hiding`’s meaning conditional on the presence of `using` seems worse.

As alternative syntax, I have thought of `open (import Foo hiding (bar))`, but that suggests (to me) that `import Foo hiding (bar)` is more first-class than it is.

**Comment by andreasabel** [comment]
I agree that it is confusing that the import directives such as `hiding` and `using` apply to either the `import` or the `open`, depending on the context.  The radical, not backwards-compatible solution would be to deprecate `import M hiding/using`.  Your suggestion (#3879), to make `import M hiding/using` first class, would be another solution.

Maybe the error message `Hiding ... has no effect` should be complemented by a remark that `hiding` applies to `open` rather than `import` in this case.

**Comment by andreasabel** [comment]
Imho, the handling of `import` is a bit too imperative.  Note that Agda accepts the following,
```agda
import Agda.Builtin.Nat using ()
open import Agda.Builtin.Nat using ()

test = Agda.Builtin.Nat.Nat
```
but if we swap the `import` statements, Agda refuses,
```agda
open import Agda.Builtin.Nat using ()
import Agda.Builtin.Nat using ()

test = Agda.Builtin.Nat.Nat
```
with error
```
Not in scope:
  Agda.Builtin.Nat.Nat
```
It would be clearer if Agda only allowed a single `import` of a module `M`, unless it gets a new name as in `import M as M'`.  Or at least disallows contradictory imports such as in my example.

**Comment by andreasabel** [comment]
I can make the situation more absurd as follows:
```agda
open import Agda.Builtin.Nat using ()

Nat  = Agda.Builtin.Nat.Nat
zero = Agda.Builtin.Nat.zero

import Agda.Builtin.Nat using ()

works : Nat
works = zero

fails : Agda.Builtin.Nat.Nat
fails = Agda.Builtin.Nat.zero
```
For the type signature of `fails`,  Agda complains:
```
Not in scope:
  Agda.Builtin.Nat.Nat
```
ROFL :D
